Register prims as operators + binding fixes

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #77117

This makes prims look as if they were defined in native_functions.yaml
but they're still all written in Python.  You now need to give a full
schema string for your prims.  The returned prim object is now
torch.ops.prim overload (prims are not allowed to be overloaded,
so we return the overload, not the overload packet, for speed.)

When I switched prims to directly call here, I had to fix the dtype/layout/memory_format `__torch_dispatch__` problem. This problem is the one where all arguments that take these enums receive integers inside `__torch_dispatch__`. I was originally intending only to update the to/from IValue logic to handle this, but it turns out C++ FunctionSchema *completely* loses this information (you ask for the type and it says int.) So I extended Argument to record the "real" type (adding MemoryFormatType as needed) and updated the schema parser to also record the real type. We then query that to do conversions to the appropriate Python objects.

I also had to fix a segfault bug for complex scalar to Python, now tested in `test_mul_complex`. I also fixed a bug where Python operator registration API did not properly propagate exceptions.

Also, I register to "prims" not "prim" because prim is already taken by TS interpreter. This is very confusing and we should find a different name.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Register prims as operators.

This makes prims look as if they were defined in native_functions.yaml
but they're still all written in Python.  You now need to give a full
schema string for your prims.  The returned prim object is now
torch.ops.prim overload (prims are not allowed to be overloaded,
so we return the overload, not the overload packet, for speed.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ngimel commented:
Hm, does that mean that all scalar args have to be wrapped before going to prims? I don't remember what state we are in, it changes too frequently.

ezyang commented:
> Hm, does that mean that all scalar args have to be wrapped before going to prims? I don't remember what state we are in, it changes too frequently.

This should actually work, because I recently changed torch.ops to unconditionally accept numbers in place of Tensor, converting them into wrapped numbers, and then they pop back out as plain ints/floats on the way out.

ngimel commented:
Oh, that's so nice! I didn't know about it.

mruberry reviewed (approved):
Seems OK -- we should update the runbook

It's too bad schema strings can't be generated from the signatures, but I guess that would require using a type annotation to specify which tensor was mutated

ezyang commented:
Technically we have enough information because you've got the VIEW thing. Probably wouldn't be too hard to make the inspect gadget but need to standardize the type annotations

ezyang committed:
Update on "Register prims as operators."


This makes prims look as if they were defined in native_functions.yaml
but they're still all written in Python.  You now need to give a full
schema string for your prims.  The returned prim object is now
torch.ops.prim overload (prims are not allowed to be overloaded,
so we return the overload, not the overload packet, for speed.)

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Register prims as operators."


This makes prims look as if they were defined in native_functions.yaml
but they're still all written in Python.  You now need to give a full
schema string for your prims.  The returned prim object is now
torch.ops.prim overload (prims are not allowed to be overloaded,
so we return the overload, not the overload packet, for speed.)

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
Reviewers this PR got substantially more involved because I ended up fixing a bunch of bugs, please take an extra look

ezyang committed:
Update on "Register prims as operators + binding fixes"


This makes prims look as if they were defined in native_functions.yaml
but they're still all written in Python.  You now need to give a full
schema string for your prims.  The returned prim object is now
torch.ops.prim overload (prims are not allowed to be overloaded,
so we return the overload, not the overload packet, for speed.)

When I switched prims to directly call here, I had to fix the dtype/layout/memory_format `__torch_dispatch__` problem. This problem is the one where all arguments that take these enums receive integers inside `__torch_dispatch__`. I was originally intending only to update the to/from IValue logic to handle this, but it turns out C++ FunctionSchema *completely* loses this information (you ask for the type and it says int.) So I extended Argument to record the "real" type (adding MemoryFormatType as needed) and updated the schema parser to also record the real type. We then query that to do conversions to the appropriate Python objects.

I also had to fix a segfault bug for complex scalar to Python, now tested in `test_mul_complex`. I also fixed a bug where Python operator registration API did not properly propagate exceptions.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Register prims as operators + binding fixes"


This makes prims look as if they were defined in native_functions.yaml
but they're still all written in Python.  You now need to give a full
schema string for your prims.  The returned prim object is now
torch.ops.prim overload (prims are not allowed to be overloaded,
so we return the overload, not the overload packet, for speed.)

When I switched prims to directly call here, I had to fix the dtype/layout/memory_format `__torch_dispatch__` problem. This problem is the one where all arguments that take these enums receive integers inside `__torch_dispatch__`. I was originally intending only to update the to/from IValue logic to handle this, but it turns out C++ FunctionSchema *completely* loses this information (you ask for the type and it says int.) So I extended Argument to record the "real" type (adding MemoryFormatType as needed) and updated the schema parser to also record the real type. We then query that to do conversions to the appropriate Python objects.

I also had to fix a segfault bug for complex scalar to Python, now tested in `test_mul_complex`. I also fixed a bug where Python operator registration API did not properly propagate exceptions.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ngimel commented on torch/csrc/jit/frontend/function_schema_parser.cpp:
```diff
@@ -148,16 +148,18 @@ struct SchemaParser {
-    auto p = type_parser.parseType();
-    auto type = std::move(p.first);
-    auto alias_info = std::move(p.second);
+    auto p = type_parser.parseFakeAndRealType();
```
can you leave a comment here or in parseFakeAndRealType what does fake type do?

ezyang committed:
Update on "Register prims as operators + binding fixes"


This makes prims look as if they were defined in native_functions.yaml
but they're still all written in Python.  You now need to give a full
schema string for your prims.  The returned prim object is now
torch.ops.prim overload (prims are not allowed to be overloaded,
so we return the overload, not the overload packet, for speed.)

When I switched prims to directly call here, I had to fix the dtype/layout/memory_format `__torch_dispatch__` problem. This problem is the one where all arguments that take these enums receive integers inside `__torch_dispatch__`. I was originally intending only to update the to/from IValue logic to handle this, but it turns out C++ FunctionSchema *completely* loses this information (you ask for the type and it says int.) So I extended Argument to record the "real" type (adding MemoryFormatType as needed) and updated the schema parser to also record the real type. We then query that to do conversions to the appropriate Python objects.

I also had to fix a segfault bug for complex scalar to Python, now tested in `test_mul_complex`. I also fixed a bug where Python operator registration API did not properly propagate exceptions.

Also, I register to "prims" not "prim" because prim is already taken by TS interpreter. This is very confusing and we should find a different name.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Register prims as operators + binding fixes"


This makes prims look as if they were defined in native_functions.yaml
but they're still all written in Python.  You now need to give a full
schema string for your prims.  The returned prim object is now
torch.ops.prim overload (prims are not allowed to be overloaded,
so we return the overload, not the overload packet, for speed.)

When I switched prims to directly call here, I had to fix the dtype/layout/memory_format `__torch_dispatch__` problem. This problem is the one where all arguments that take these enums receive integers inside `__torch_dispatch__`. I was originally intending only to update the to/from IValue logic to handle this, but it turns out C++ FunctionSchema *completely* loses this information (you ask for the type and it says int.) So I extended Argument to record the "real" type (adding MemoryFormatType as needed) and updated the schema parser to also record the real type. We then query that to do conversions to the appropriate Python objects.

I also had to fix a segfault bug for complex scalar to Python, now tested in `test_mul_complex`. I also fixed a bug where Python operator registration API did not properly propagate exceptions.

Also, I register to "prims" not "prim" because prim is already taken by TS interpreter. This is very confusing and we should find a different name.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Register prims as operators + binding fixes"


This makes prims look as if they were defined in native_functions.yaml
but they're still all written in Python.  You now need to give a full
schema string for your prims.  The returned prim object is now
torch.ops.prim overload (prims are not allowed to be overloaded,
so we return the overload, not the overload packet, for speed.)

When I switched prims to directly call here, I had to fix the dtype/layout/memory_format `__torch_dispatch__` problem. This problem is the one where all arguments that take these enums receive integers inside `__torch_dispatch__`. I was originally intending only to update the to/from IValue logic to handle this, but it turns out C++ FunctionSchema *completely* loses this information (you ask for the type and it says int.) So I extended Argument to record the "real" type (adding MemoryFormatType as needed) and updated the schema parser to also record the real type. We then query that to do conversions to the appropriate Python objects.

I also had to fix a segfault bug for complex scalar to Python, now tested in `test_mul_complex`. I also fixed a bug where Python operator registration API did not properly propagate exceptions.

Also, I register to "prims" not "prim" because prim is already taken by TS interpreter. This is very confusing and we should find a different name.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

albanD commented on test/test_ops.py:
```diff
@@ -151,6 +151,8 @@ def unsupported(dtype):
                     supported_dtypes.add(dtype)
                 except Exception as e:
+                    import traceback
+                    traceback.print_exc()
```
Debug print?

albanD commented on test/test_prims.py:
```diff
@@ -77,6 +78,23 @@ def _wrapper(a, shape, broadcast_dimensions):
+            prims.sin(r)
+        self.assertExpectedInline('\n'.join(logs), """\
+$0 = input('input')
+$1 = torch._ops.prims.sin.default($0)""")
```
We should really fix the `__module__` of the torch.ops functions

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -99,17 +99,34 @@ std::pair<py::object, py::dict> parseIValuesToPyArgsKwargs(const c10::OperatorHa
+    } else if (arg.real_type()->kind() == c10::MemoryFormatType::Kind) {
+      // TODO: https://github.com/pytorch/pytorch/issues/77135
+      auto* obj = THPMemoryFormat_New(static_cast<c10::MemoryFormat>(arguments[idx].toInt()), "unused");
+      return py::reinterpret_borrow<py::object>(reinterpret_cast<PyObject*>(obj));
```
THPMemoryFormat_New returns a new reference so this should be a steal no?

albanD reviewed (commented):
Sounds mostly good to me.
Very nice cleanup!

ezyang committed:
Update on "Register prims as operators + binding fixes"


This makes prims look as if they were defined in native_functions.yaml
but they're still all written in Python.  You now need to give a full
schema string for your prims.  The returned prim object is now
torch.ops.prim overload (prims are not allowed to be overloaded,
so we return the overload, not the overload packet, for speed.)

When I switched prims to directly call here, I had to fix the dtype/layout/memory_format `__torch_dispatch__` problem. This problem is the one where all arguments that take these enums receive integers inside `__torch_dispatch__`. I was originally intending only to update the to/from IValue logic to handle this, but it turns out C++ FunctionSchema *completely* loses this information (you ask for the type and it says int.) So I extended Argument to record the "real" type (adding MemoryFormatType as needed) and updated the schema parser to also record the real type. We then query that to do conversions to the appropriate Python objects.

I also had to fix a segfault bug for complex scalar to Python, now tested in `test_mul_complex`. I also fixed a bug where Python operator registration API did not properly propagate exceptions.

Also, I register to "prims" not "prim" because prim is already taken by TS interpreter. This is very confusing and we should find a different name.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Register prims as operators + binding fixes"


This makes prims look as if they were defined in native_functions.yaml
but they're still all written in Python.  You now need to give a full
schema string for your prims.  The returned prim object is now
torch.ops.prim overload (prims are not allowed to be overloaded,
so we return the overload, not the overload packet, for speed.)

When I switched prims to directly call here, I had to fix the dtype/layout/memory_format `__torch_dispatch__` problem. This problem is the one where all arguments that take these enums receive integers inside `__torch_dispatch__`. I was originally intending only to update the to/from IValue logic to handle this, but it turns out C++ FunctionSchema *completely* loses this information (you ask for the type and it says int.) So I extended Argument to record the "real" type (adding MemoryFormatType as needed) and updated the schema parser to also record the real type. We then query that to do conversions to the appropriate Python objects.

I also had to fix a segfault bug for complex scalar to Python, now tested in `test_mul_complex`. I also fixed a bug where Python operator registration API did not properly propagate exceptions.

Also, I register to "prims" not "prim" because prim is already taken by TS interpreter. This is very confusing and we should find a different name.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Register prims as operators + binding fixes"


This makes prims look as if they were defined in native_functions.yaml
but they're still all written in Python.  You now need to give a full
schema string for your prims.  The returned prim object is now
torch.ops.prim overload (prims are not allowed to be overloaded,
so we return the overload, not the overload packet, for speed.)

When I switched prims to directly call here, I had to fix the dtype/layout/memory_format `__torch_dispatch__` problem. This problem is the one where all arguments that take these enums receive integers inside `__torch_dispatch__`. I was originally intending only to update the to/from IValue logic to handle this, but it turns out C++ FunctionSchema *completely* loses this information (you ask for the type and it says int.) So I extended Argument to record the "real" type (adding MemoryFormatType as needed) and updated the schema parser to also record the real type. We then query that to do conversions to the appropriate Python objects.

I also had to fix a segfault bug for complex scalar to Python, now tested in `test_mul_complex`. I also fixed a bug where Python operator registration API did not properly propagate exceptions.

Also, I register to "prims" not "prim" because prim is already taken by TS interpreter. This is very confusing and we should find a different name.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Register prims as operators + binding fixes"


This makes prims look as if they were defined in native_functions.yaml
but they're still all written in Python.  You now need to give a full
schema string for your prims.  The returned prim object is now
torch.ops.prim overload (prims are not allowed to be overloaded,
so we return the overload, not the overload packet, for speed.)

When I switched prims to directly call here, I had to fix the dtype/layout/memory_format `__torch_dispatch__` problem. This problem is the one where all arguments that take these enums receive integers inside `__torch_dispatch__`. I was originally intending only to update the to/from IValue logic to handle this, but it turns out C++ FunctionSchema *completely* loses this information (you ask for the type and it says int.) So I extended Argument to record the "real" type (adding MemoryFormatType as needed) and updated the schema parser to also record the real type. We then query that to do conversions to the appropriate Python objects.

I also had to fix a segfault bug for complex scalar to Python, now tested in `test_mul_complex`. I also fixed a bug where Python operator registration API did not properly propagate exceptions.

Also, I register to "prims" not "prim" because prim is already taken by TS interpreter. This is very confusing and we should find a different name.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

albanD reviewed (approved):
SGTM

ezyang commented:
@pytorchbot merge this

ezyang committed:
Update on "Register prims as operators + binding fixes"


This makes prims look as if they were defined in native_functions.yaml
but they're still all written in Python.  You now need to give a full
schema string for your prims.  The returned prim object is now
torch.ops.prim overload (prims are not allowed to be overloaded,
so we return the overload, not the overload packet, for speed.)

When I switched prims to directly call here, I had to fix the dtype/layout/memory_format `__torch_dispatch__` problem. This problem is the one where all arguments that take these enums receive integers inside `__torch_dispatch__`. I was originally intending only to update the to/from IValue logic to handle this, but it turns out C++ FunctionSchema *completely* loses this information (you ask for the type and it says int.) So I extended Argument to record the "real" type (adding MemoryFormatType as needed) and updated the schema parser to also record the real type. We then query that to do conversions to the appropriate Python objects.

I also had to fix a segfault bug for complex scalar to Python, now tested in `test_mul_complex`. I also fixed a bug where Python operator registration API did not properly propagate exceptions.

Also, I register to "prims" not "prim" because prim is already taken by TS interpreter. This is very confusing and we should find a different name.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge on green
